### PR TITLE
#HT-88: Allow execution of MySQL object queries in the `query-engine`

### PR DESCRIPTION
**Pull request by joelerego**

### CODE REVIEW

File: packages/query-engine/source/app.ts [hunk 1]
```diff
@@ -4,16 +4,23 @@ import logger from "morgan";
...
 
+    // Attach router
     const router = express.Router();
     queryEngine.attachRoutes(router);
     app.use("/api/v1", router);
 
+    // Attach error handlers
+    app.use(BadRequestError);
+    app.use(NotFoundError);
+    app.use(UnauthorizedError);
```
Comment (itssamuelrowe) [inline packages/query-engine/source/app.ts:22]: 1. In a way, it makes sense to split the error handling. But then again, why not combine all the error handlers into a single error handler? I'm in favor of combing all the error handlers together.
2. Why are you using PascalCase for middleware? Further, people will get confused with exception classes from `@hypertool/common` with the same name.

File: packages/query-engine/source/app.ts [hunk 2]
```diff
@@ -4,16 +4,23 @@ import logger from "morgan";
 import cors from "cors";
 
 import { queryEngine } from "./rest";
+import { BadRequestError, NotFoundError, UnauthorizedError } from "./utils";
```
Comment (itssamuelrowe) [inline packages/query-engine/source/app.ts:7]: Middleware should go in the `middleware` directory, not utils.

File: packages/query-engine/source/controllers/query-engine-controller.ts [hunk 3]
```diff
@@ -1,47 +1,86 @@
...
 
-import mysql from "mysql2/promise";
 import { QueryTemplateModel, ResourceModel } from "@hypertool/common";
-import { NextFunction } from "express";
+import { Knex, knex } from "knex";
 
 import type { ExecuteParameters } from "../types";
 
-const execute = async (
+const executeMySQL = async (
```
Comment (itssamuelrowe) [inline packages/query-engine/source/controllers/query-engine-controller.ts:8]: Since we are using Knex, why do you need a specific function for MySQL? Is there a way to make this function generic?

File: packages/query-engine/source/controllers/query-engine-controller.ts [hunk 4]
```diff
@@ -1,47 +1,86 @@
...
 
-    const resource: Resource = await ResourceModel.findOne({
-        _id: query.resource,
-    }).exec();
+    if (typeof query.content === "object") {
+        const queryContent: any = query.content;
 
-    const mySQLConfig: MySQLConfiguration = resource.mysql;
+        let queryResult;
+        if (queryContent.hasOwnProperty("select") && queryContent.select) {
```
Comment (itssamuelrowe) [inline packages/query-engine/source/controllers/query-engine-controller.ts:30]: The query building needs to happen in a tree traversal fashion, recursively. For example, the first stage handles what kind of operation: select, insert, delete, and so on. The second stage takes care of from clause. The third stage takes care of where clause.

Why?

1. Duplicate code will be eliminated.
2. The objects can become more complicated. For example, in the current design, you cannot have optional where clauses.
3. The code will be concise, neat, and of better quality.

File: packages/query-engine/source/controllers/query-engine-controller.ts [hunk 5]
```diff
@@ -1,47 +1,86 @@
...
+    try {
+        const query: Query = await QueryTemplateModel.findOne({
+            name: queryRequest.name,
+        }).exec();
+
+        const resource: Resource = await ResourceModel.findOne({
+            _id: query.resource,
+        }).exec();
+
+        console.log(query);
```
Comment (itssamuelrowe) [inline packages/query-engine/source/controllers/query-engine-controller.ts:69]: Remove this.

File: packages/query-engine/source/controllers/query-engine-controller.ts [hunk 6]
```diff
@@ -1,47 +1,86 @@
...
+
+        switch (resource.type) {
+            case "mysql": {
```
Comment (itssamuelrowe) [inline packages/query-engine/source/controllers/query-engine-controller.ts:72]: We can include other SQL resource types that are inherently supported by Knex here.

File: packages/query-engine/source/controllers/query-engine-controller.ts [hunk 7]
```diff
@@ -1,47 +1,86 @@
...
+                const queryResult = await executeMySQL(
+                    queryRequest,
+                    query,
+                    resource,
+                );
+                return { result: queryResult };
```
Comment (itssamuelrowe) [inline packages/query-engine/source/controllers/query-engine-controller.ts:78]: Isn't the data supposed to be formated?

File: packages/query-engine/source/utils/errors/BadRequestError.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+const BadRequestError = (
+    error: any,
+    request: Request,
+    response: Response,
+    next: NextFunction,
+): void => {
+    if (error.code === "ER_BAD_HOST_ERROR") {
+        response.status(httpStatuses.BAD_REQUEST).json({
+            error: "Bad Host Request",
```
Comment (itssamuelrowe) [inline packages/query-engine/source/utils/errors/BadRequestError.ts:14]: Error messages should be proper sentences. That's the convention we follow everywhere else.

File: packages/query-engine/source/utils/errors/NotFoundError.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+const { httpStatuses } = constants;
+
+const NotFoundError = (
+    error: any,
+    request: Request,
+    response: Response,
+    next: NextFunction,
+): void => {
+    if (
+        error.code === "ER_BAD_DB_ERROR" ||
```
Comment (itssamuelrowe) [inline packages/query-engine/source/utils/errors/NotFoundError.ts:13]: Are these error codes from Knex? Or did you create them? If you created them, please no abbreviations.

**Review by itssamuelrowe — CHANGES_REQUESTED** [review]

File: packages/query-engine/source/utils/errors/NotFoundError.ts (hunk 9 above)
Comment (joelerego) [reply packages/query-engine/source/utils/errors/NotFoundError.ts:13]: Error codes are from SQL. They are SQL-specific.

File: packages/query-engine/source/utils/errors/BadRequestError.ts (hunk 8 above)
Comment (joelerego) [reply packages/query-engine/source/utils/errors/BadRequestError.ts:14]: The messages are full sentences here. The `error` field is the name of the error.

File: packages/query-engine/source/controllers/query-engine-controller.ts (hunk 4 above)
Comment (joelerego) [reply packages/query-engine/source/controllers/query-engine-controller.ts:30]: Agreed.

Comment (joelerego) [reply packages/query-engine/source/controllers/query-engine-controller.ts:31]: We can have a query building login that uses tree traversal fashion, but non-recursive for now. Is that okay?

File: packages/query-engine/source/controllers/query-engine-controller.ts [hunk 10]
```diff
@@ -1,47 +1,71 @@
...
-            typeof queryRequest.variables === "object" ? true : false,
-    });
-
-    let queryResult: any;
-    try {
-        const [results, fields] = await connection.execute(
+    if (typeof query.content === "object") {
+        const queryBuilder = new QueryBuilder(knexInstance);
+        queryBuilder.parse(query.content);
+        const queryResult = await queryBuilder.knexInstance();
```
Comment (itssamuelrowe) [inline packages/query-engine/source/controllers/query-engine-controller.ts:31]: 1. You already have access to `knexInstance`. So why are you accessing it via `queryBuilder`?
2. Rename `knexInstance` to something better. `instance` should be enough I think.

File: packages/query-engine/source/controllers/query-engine-controller.ts [hunk 11]
```diff
@@ -1,47 +1,71 @@
...
         );
-        queryResult = { results, fields };
-    } catch (error) {
-        next(error);
+        return queryResult[0];
     }
+};
 
-    return queryResult;
+const execute = async (queryRequest: ExecuteParameters): Promise<any> => {
```
Comment (itssamuelrowe) [inline packages/query-engine/source/controllers/query-engine-controller.ts:42]: `parameters`

File: packages/query-engine/source/controllers/query-engine-controller.ts [hunk 12]
```diff
@@ -1,47 +1,71 @@
...
+    try {
+        const query: Query = await QueryTemplateModel.findOne({
+            name: queryRequest.name,
+        }).exec();
+
+        const resource: Resource = await ResourceModel.findOne({
+            _id: query.resource,
+        }).exec();
```
Comment (itssamuelrowe) [inline packages/query-engine/source/controllers/query-engine-controller.ts:50]: You can combine both these statements using `populate`.

`MyModel.findOne({ ... }).populate("myRefField")`

File: packages/query-engine/source/controllers/query-engine-controller.ts [hunk 13]
```diff
@@ -1,47 +1,71 @@
...
+                );
+                return { result: queryResult };
+            }
+            default: {
+                throw new Error("Unknown resource type");
+            }
+        }
+    } catch (error) {
+        throw error;
+    }
```
Comment (itssamuelrowe) [inline packages/query-engine/source/controllers/query-engine-controller.ts:68]: What's the point of catching the error?

File: packages/query-engine/source/middleware/errors/bad-request-error.ts [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+const BadRequestError = (
+    error: any,
+    request: Request,
+    response: Response,
+    next: NextFunction,
+): void => {
+    if (error.code === "ER_BAD_HOST_ERROR") {
+        response.status(httpStatuses.BAD_REQUEST).json({
+            error: "Bad Host Request",
```
Comment (itssamuelrowe) [inline packages/query-engine/source/middleware/errors/bad-request-error.ts:14]: Is `error` supposed to be a code that the client will understand? If yes, don't include spaces. I'd rather prefer `BAD_HOST_REQUEST`.

File: packages/query-engine/source/middleware/errors/not-found-error.ts [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+    response: Response,
+    next: NextFunction,
+): void => {
+    if (
+        error.code === "ER_BAD_DB_ERROR" ||
+        error.code === "ER_BAD_TABLE_ERROR" ||
+        error.code === "ER_NO_SUCH_TABLE"
+    ) {
+        response.status(httpStatuses.NOT_FOUND).json({
+            error: "Not Found",
```
Comment (itssamuelrowe) [inline packages/query-engine/source/middleware/errors/not-found-error.ts:18]: If this is an error code, `NOT_FOUND` is better. Also, wouldn't it be better if the user had more information than just "NOT_FOUND"? Let's retain the error codes. Create a mapping in constants.

ER_BAD_DB_ERROR -> BAD_DB_ERROR
ER_BAD_TABLE_ERROR -> BAD_TABLE_ERROR
NO_SUCH_TABLE -> NO_SUCH_TABLE_ERROR

and the other errors

File: packages/query-engine/source/middleware/errors/unauthorized-error.ts [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+const UnauthorizedError = (
+    error: any,
+    request: Request,
+    response: Response,
+    next: NextFunction,
+): void => {
+    if (error.code === "ER_ACCESS_DENIED_ERROR") {
+        response.status(httpStatuses.UNAUTHORIZED).json({
+            error: "Unauthorized",
```
Comment (itssamuelrowe) [inline packages/query-engine/source/middleware/errors/unauthorized-error.ts:14]: ...

File: packages/query-engine/source/utils/query-builder.ts [hunk 17]
```diff
@@ -0,0 +1,83 @@
+import { Knex, knex } from "knex";
+
+export class QueryBuilder {
+    knexInstance: Knex;
```
Comment (itssamuelrowe) [inline packages/query-engine/source/utils/query-builder.ts:4]: `instance`

File: packages/query-engine/source/utils/query-builder.ts [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+     * }
+     *
+     * constructor(knexInstance: Knex, tableName: string, alias: string) {
+     *   this.knexInstance = knexInstance.table(tableName).as(alias);
+     * }
+     *
+     * constructor(config: Knex.Config) {
+     *   this.knexInstance = knex(config);
+     * }
+     */
```
Comment (itssamuelrowe) [inline packages/query-engine/source/utils/query-builder.ts:25]: Does TypeScript support constructor overloading?

File: packages/query-engine/source/utils/query-builder.ts [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+
+    /**
+     * `parse()`
```
Comment (itssamuelrowe) [inline packages/query-engine/source/utils/query-builder.ts:28]: Why is this required?

**Review by itssamuelrowe — CHANGES_REQUESTED** [review]

File: packages/query-engine/source/utils/query-builder.ts (hunk 18 above)
Comment (joelerego) [reply packages/query-engine/source/utils/query-builder.ts]: TypeScript allows you to declare overloads but you can only have one implementation and that implementation must have a signature that is compatible with all overloads. If that is too messy, we can use the JS pattern with one constructor with switch statements based on args.

**Review by itssamuelrowe — APPROVED** [review]
